interceptor/reflection.go

```diff
@@ -220,21 +224,55 @@ func translateDataBlobs(match stringMatcher, visitor visitor, blobs ...*common.D
 	return blobs, anyChanged, nil
 }
 
-func translateOneDataBlob(match stringMatcher, visitor visitor, blob *common.DataBlob) (*common.DataBlob, bool, error) {
+func translateOneDataBlob(logger log.Logger, match stringMatcher, visitor visitor, blob *common.DataBlob) (*common.DataBlob, bool, error) {
 	if blob == nil || len(blob.Data) == 0 {
 		return blob, false, nil
 
 	}
-	evt, err := serializer.DeserializeEvents(blob)
+	events, err := serializer.DeserializeEvents(blob)
 	if err != nil {
 		return blob, false, err
 	}
 
-	changed, err := visitor(evt, match)
-	if err != nil || !changed {
-		return blob, changed, err
+	var anyChanged bool
+	changed, err := validateAndRepairHistoryEvents(logger, events)
+	anyChanged = anyChanged || changed
+	if err != nil {
+		return blob, anyChanged, err
+	}
+
+	changed, err = visitor(logger, events, match)
+	anyChanged = anyChanged || changed
+	if err != nil {
+		return blob, anyChanged, err
+	}
+
+	if !anyChanged {
+		// skip reserializing if there are no changes
+		return blob, anyChanged, err
 	}
 
-	newBlob, err := serializer.SerializeEvents(evt, blob.EncodingType)
-	return newBlob, changed, err
+	newBlob, err := serializer.SerializeEvents(events, blob.EncodingType)
+	return newBlob, anyChanged, err
+}
+
+// In old versions of Temporal, it was possible that certain history events could
+// be written with invalid UTF-8.
+func validateAndRepairHistoryEvents(logger log.Logger, events []*history.HistoryEvent) (bool, error) {
+	var changed bool
+	for _, event := range events {
+		// only this field has been observed to have bad data
+		cause := event.
+			GetActivityTaskStartedEventAttributes().
+			GetLastFailure().
+			GetCause()
+		if !utf8.ValidString(cause.GetMessage()) {
+			changed = true
+			cause.Message = strings.ToValidUTF8(cause.Message, string(utf8.RuneError))
+
+			logger.Warn("repaired invalid utf-8 in history event")
```

Comment:
not sure if we can log workflow id/run id so we can compare.

Reply:
I looked but didn't find a way to get the workflow id or run id easily. We would need a clever way to discover it I think